ui/ozone/platform/wayland/wayland_input_method_context.cc

```diff
@@ -0,0 +1,126 @@
+}
+
+WaylandInputMethodContext::~WaylandInputMethodContext() {
+  if (text_input_) {
+    text_input_->Deactivate();
+    text_input_->HideInputPanel();
+  }
+}
+
+void WaylandInputMethodContext::Initialize(
+    ui::mojom::LinuxInputMethodContextDelegatePtr delegate,
+    bool is_simple) {
+  delegate_ = std::move(delegate);
+  if (use_ozone_wayland_vkb_ && !is_simple &&
+      connection_->text_input_manager_v1()) {
+    text_input_ = std::make_unique<ZWPTextInputWrapperV1>(
+        connection_->text_input_manager_v1());
+    text_input_->Initialize(connection_, this);
+  }
+}
+
+void WaylandInputMethodContext::DispatchKeyEvent(
+    std::unique_ptr<ui::Event> key_event,
+    DispatchKeyEventCallback callback) {
+  const bool handled = false;
+  std::move(callback).Run(handled);
+}
+
+void WaylandInputMethodContext::Reset() {
+  if (text_input_)
+    text_input_->Reset();
+}
+
+void WaylandInputMethodContext::Focus() {
+  if (!text_input_ || !connection_->GetCurrentKeyboardFocusedWindow())
+    return;
+
+  text_input_->Activate(connection_->GetCurrentKeyboardFocusedWindow());
+  text_input_->ShowInputPanel();
+}
+
+void WaylandInputMethodContext::Blur() {
+  if (text_input_) {
+    text_input_->Deactivate();
+    text_input_->HideInputPanel();
+  }
+}
+
+void WaylandInputMethodContext::SetCursorLocation(const gfx::Rect& rect) {
+  if (text_input_)
+    text_input_->SetCursorRect(rect);
+}
+
+void WaylandInputMethodContext::OnPreeditString(const std::string& text,
+                                                int preedit_cursor) {
+  gfx::Range selection_range = gfx::Range::InvalidRange();
+
+  // TODO(jani) Handle selection range
+
+  if (!selection_range.IsValid()) {
+    int cursor_pos = (preedit_cursor) ? text.length() : preedit_cursor;
+    selection_range.set_start(cursor_pos);
+    selection_range.set_end(cursor_pos);
+  }
+
+  ui::CompositionText composition_text;
+  composition_text.text = base::UTF8ToUTF16(text);
+  composition_text.selection = selection_range;
+  delegate_->OnPreeditChanged(composition_text);
+}
+
+void WaylandInputMethodContext::OnCommitString(const std::string& text) {
+  delegate_->OnCommit(base::UTF8ToUTF16(text));
+}
+
+void WaylandInputMethodContext::OnKeysym(uint32_t key,
+                                         uint32_t state,
+                                         uint32_t modifiers) {
+  uint8_t flags = 0;  // for now ignore modifiers
+  DomKey dom_key = NonPrintableXKeySymToDomKey(key);
+  KeyboardCode key_code = NonPrintableDomKeyToKeyboardCode(dom_key);
+  DomCode dom_code =
+      KeycodeConverter::NativeKeycodeToDomCode(key_code + kXkbKeycodeOffset);
+  if (dom_code == ui::DomCode::NONE)
+    return;
+
+  bool down = state == WL_KEYBOARD_KEY_STATE_PRESSED;
+  ui::KeyEvent event(down ? ET_KEY_PRESSED : ET_KEY_RELEASED, key_code,
+                     dom_code, flags, dom_key, EventTimeForNow());
+  connection_->DispatchUiEvent(&event);
```

Comment:
rather than calling WaylandConnection::DispatchUiEvent directly, can not we pass a Callback, like we do in WaylandKeyboard or WaylandPointer?


291       connection->keyboard_ = std::make_unique<WaylandKeyboard>(
292           keyboard, base::Bind(&WaylandConnection::DispatchUiEvent,
293                                base::Unretained(connection)));

Reply:
But in that case WaylandInputMethodContextTest and OzonePlatformWayland needs to be declared as friend classes for WaylandConnection, right? Is that better approach?

Reply:
Hum, this is a good point. If other tests do this, it should be fine for us too.